fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/FilterWithConjunctNestedToNestingContextRule.java

```diff
@@ -0,0 +1,94 @@
+import com.apple.foundationdb.record.query.plan.temp.PlannerRuleCall;
+import com.apple.foundationdb.record.query.plan.temp.expressions.LogicalFilterExpression;
+import com.apple.foundationdb.record.query.plan.temp.expressions.NestedContextExpression;
+import com.apple.foundationdb.record.query.plan.temp.expressions.RelationalPlannerExpression;
+import com.apple.foundationdb.record.query.plan.temp.matchers.AnyChildWithRestMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ExpressionMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ReferenceMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.TypeMatcher;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * A rule that looks for a logical filter that contains a conjunction that includes a non-repeated nested predicate
+ * ({@link NestedField}) and transforms the inner expression to a version with respect to the {@link NestedContext}
+ * defined by that nested field. The other filters in the conjunct are left above the produced
+ * {@link NestedContextExpression} and need to be handled separately.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class FilterWithConjunctNestedToNestingContextRule extends PlannerRule<LogicalFilterExpression> {
+    private static final ExpressionMatcher<ExpressionRef<RelationalPlannerExpression>> innerMatcher = ReferenceMatcher.anyRef();
+    private static final ExpressionMatcher<ExpressionRef<QueryComponent>> nestedFilterMatcher = ReferenceMatcher.anyRef();
+    private static final ExpressionMatcher<NestedField> nestedFieldMatcher = TypeMatcher.of(NestedField.class, nestedFilterMatcher);
+    private static final ReferenceMatcher<QueryComponent> otherConjunctMatcher = ReferenceMatcher.anyRef();
+    private static final ExpressionMatcher<LogicalFilterExpression> root = TypeMatcher.of(LogicalFilterExpression.class,
+            TypeMatcher.of(AndComponent.class, AnyChildWithRestMatcher.anyMatchingWithRest(
+                    nestedFieldMatcher, otherConjunctMatcher)), innerMatcher);
+
+    public FilterWithConjunctNestedToNestingContextRule() {
+        super(root);
+    }
+
+    @Nonnull
+    @Override
+    public ChangesMade onMatch(@Nonnull PlannerRuleCall call) {
+        final ExpressionRef<RelationalPlannerExpression> inner = call.get(innerMatcher);
+        final NestedField nestedField = call.get(nestedFieldMatcher);
+        final List<ExpressionRef<QueryComponent>> otherConjuncts = call.getBindings().getAll(otherConjunctMatcher);
+        final ExpressionRef<QueryComponent> nestedFilter = call.get(nestedFilterMatcher);
+        call.yield(call.ref(call.get(root)));
+
+        final NestedContext nestedContext = new SingleNestedContext(nestedField.getFieldName());
+        final ExpressionRef<RelationalPlannerExpression> nestedInner = nestedContext.getNestedRelationalPlannerExpression(inner);
+
+        if (nestedInner == null) {
+            return ChangesMade.NO_CHANGE;
```

Comment:
It looks like this rule returns `NO_CHANGE` when it only yields back the `root`, which I think is consistent with the rules in the final commit, but not necessarily the other commits.